File: telemetry/src/main/native/cpp/MultiTelemetryBackend.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "wpi/telemetry/MultiTelemetryBackend.hpp"

#include <atomic>
#include <memory>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

#include "wpi/telemetry/TelemetryEntry.hpp"

using namespace wpi;
using namespace wpi::telemetry;

namespace {
void ValidateBackends(
    std::span<const std::shared_ptr<TelemetryBackend>> backends) {
  for (const auto& backend : backends) {
    if (!backend) {
      throw std::invalid_argument{
          "MultiTelemetryBackend child backend cannot be null"};
    }
  }
}
}  // namespace

class MultiTelemetryBackend::Entry : public TelemetryEntry {
 public:
  Entry(std::string_view path,
        const std::vector<std::shared_ptr<TelemetryBackend>>& backends) {
    m_entries.reserve(backends.size());
    for (const auto& backend : backends) {
      m_entries.emplace_back(backend->GetEntry(path));
    }
  }

  void MarkRemoved() { m_removed.store(true); }

  bool IsDiscard() const override {
    if (m_removed.load()) {
      return true;
    }
    for (const auto& entry : m_entries) {
      if (!entry->IsDiscard()) {
        return false;
      }
    }
    return true;
  }

  void KeepDuplicates() override {
    ForEachEntry([](TelemetryEntry& entry) { entry.KeepDuplicates(); });
  }

  void SetProperty(std::string_view key, std::string_view value) override {
    ForEachEntry([&](TelemetryEntry& entry) { entry.SetProperty(key, value); });
  }

  void LogBoolean(bool value) override {
    ForEachEntry([=](TelemetryEntry& entry) { entry.LogBoolean(value); });
  }

  void LogInt8(int8_t value) override {
    ForEachEntry([=](TelemetryEntry& entry) { entry.LogInt8(value); });
  }

  void LogInt16(int16_t value) override {
    ForEachEntry([=](TelemetryEntry& entry) { entry.LogInt16(value); });
  }

  void LogInt32(int32_t value) override {
    ForEachEntry([=](TelemetryEntry& entry) { entry.LogInt32(value); });
  }

  void LogInt64(int64_t value) override {
    ForEachEntry([=](TelemetryEntry& entry) { entry.LogInt64(value); });
  }

  void LogFloat(float value) override {
    ForEachEntry([=](TelemetryEntry& entry) { entry.LogFloat(value); });
  }

  void LogDouble(double value) override {
    ForEachEntry([=](TelemetryEntry& entry) { entry.LogDouble(value); });
  }

  void LogString(std::string_view value, std::string_view typeString) override {
    ForEachEntry(
        [=](TelemetryEntry& entry) { entry.LogString(value, typeString); });
  }

  void LogBooleanArray(std::span<const bool> value) override {
    ForEachEntry([=](TelemetryEntry& entry) { entry.LogBooleanArray(value); });
  }

  void LogBooleanArray(std::span<const int> value) override {
    ForEachEntry([=](TelemetryEntry& entry) { entry.LogBooleanArray(value); });
  }

  void LogInt16Array(std::span<const int16_t> value) override {
    ForEachEntry([=](TelemetryEntry& entry) { entry.LogInt16Array(value); });
  }

  void LogInt32Array(std::span<const int32_t> value) override {
    ForEachEntry([=](TelemetryEntry& entry) { entry.LogInt32Array(value); });
  }

  void LogInt64Array(std::span<const int64_t> value) override {
    ForEachEntry([=](TelemetryEntry& entry) { entry.LogInt64Array(value); });
  }

  void LogFloatArray(std::span<const float> value) override {
    ForEachEntry([=](TelemetryEntry& entry) { entry.LogFloatArray(value); });
  }

  void LogDoubleArray(std::span<const double> value) override {
    ForEachEntry([=](TelemetryEntry& entry) { entry.LogDoubleArray(value); });
  }

  void LogStringArray(std::span<const std::string> value) override {
    ForEachEntry([=](TelemetryEntry& entry) { entry.LogStringArray(value); });
  }

  void LogStringArray(std::span<const std::string_view> value) override {
    ForEachEntry([=](TelemetryEntry& entry) { entry.LogStringArray(value); });
  }

  void LogRaw(std::span<const uint8_t> value,
              std::string_view typeString) override {
    ForEachEntry(
        [=](TelemetryEntry& entry) { entry.LogRaw(value, typeString); });
  }

 private:
  template <typename F>
  void ForEachEntry(F&& func) {
    if (m_removed.load()) {
      return;
    }
    for (const auto& entry : m_entries) {
      if (m_removed.load()) {
        return;
      }
      func(*entry);
    }
  }

  std::vector<std::shared_ptr<TelemetryEntry>> m_entries;
  std::atomic_bool m_removed{false};
};

MultiTelemetryBackend::MultiTelemetryBackend() = default;

MultiTelemetryBackend::MultiTelemetryBackend(
    std::span<const std::shared_ptr<TelemetryBackend>> backends)
    : m_backends{backends.begin(), backends.end()} {
  ValidateBackends(
      std::span<const std::shared_ptr<TelemetryBackend>>{m_backends});
}

MultiTelemetryBackend::MultiTelemetryBackend(
    std::vector<std::shared_ptr<TelemetryBackend>> backends)
    : m_backends{std::move(backends)} {
  ValidateBackends(
      std::span<const std::shared_ptr<TelemetryBackend>>{m_backends});
}

MultiTelemetryBackend::MultiTelemetryBackend(
    std::initializer_list<std::shared_ptr<TelemetryBackend>> backends)
    : MultiTelemetryBackend{std::span<const std::shared_ptr<TelemetryBackend>>{
          backends.begin(), backends.size()}} {}

MultiTelemetryBackend::~MultiTelemetryBackend() = default;

std::shared_ptr<TelemetryEntry> MultiTelemetryBackend::GetEntry(
    std::string_view path) {
  std::scoped_lock lock{m_mutex};
  auto it = m_entries.find(path);
  if (it != m_entries.end()) {
    return it->second;
  }

  auto entry = std::make_shared<Entry>(path, m_backends);
  m_entries.try_emplace(path, entry);
  return entry;
}

void MultiTelemetryBackend::RemoveEntry(std::string_view path) {
  {
    std::scoped_lock lock{m_mutex};
    auto it = m_entries.find(path);
    if (it != m_entries.end()) {
      auto entry = std::move(it->second);
      m_entries.erase(it);
      entry->MarkRemoved();
    }
  }
  for (const auto& backend : m_backends) {
    backend->RemoveEntry(path);
  }
}

bool MultiTelemetryBackend::HasSchema(std::string_view schemaName) const {
  for (const auto& backend : m_backends) {
    if (!backend->HasSchema(schemaName)) {
      return false;
    }
  }
  return true;
}

void MultiTelemetryBackend::AddSchema(std::string_view schemaName,
                                      std::string_view type,
                                      std::span<const uint8_t> schema) {
  for (const auto& backend : m_backends) {
    backend->AddSchema(schemaName, type, schema);
  }
}

void MultiTelemetryBackend::AddSchema(std::string_view schemaName,
                                      std::string_view type,
                                      std::string_view schema) {
  for (const auto& backend : m_backends) {
    backend->AddSchema(schemaName, type, schema);
  }
}

File: telemetry/src/main/java/org/wpilib/telemetry/MultiTelemetryBackend.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package org.wpilib.telemetry;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.IdentityHashMap;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import org.wpilib.util.protobuf.Protobuf;
import org.wpilib.util.struct.Struct;

/** A telemetry backend that forwards logged data to multiple child backends. */
public class MultiTelemetryBackend implements TelemetryBackend {
  private final List<TelemetryBackend> m_backends;
  private final ConcurrentMap<String, Entry> m_entries = new ConcurrentHashMap<>();

  /**
   * Constructs a telemetry backend that forwards to multiple child backends.
   *
   * @param backends child backends
   */
  public MultiTelemetryBackend(TelemetryBackend... backends) {
    this(Arrays.asList(Objects.requireNonNull(backends, "backends")));
  }

  /**
   * Constructs a telemetry backend that forwards to multiple child backends.
   *
   * @param backends child backends
   */
  public MultiTelemetryBackend(Collection<? extends TelemetryBackend> backends) {
    m_backends = List.copyOf(Objects.requireNonNull(backends, "backends"));
  }

  @Override
  @SuppressWarnings("PMD.CompareObjectsWithEquals")
  public boolean ownsBackend(TelemetryBackend backend) {
    if (this == backend) {
      return true;
    }
    for (TelemetryBackend child : m_backends) {
      if (child.ownsBackend(backend)) {
        return true;
      }
    }
    return false;
  }

  @Override
  public boolean sharesBackendWith(TelemetryBackend backend) {
    if (TelemetryBackend.super.sharesBackendWith(backend)) {
      return true;
    }
    for (TelemetryBackend child : m_backends) {
      if (child.sharesBackendWith(backend)) {
        return true;
      }
    }
    return false;
  }

  @Override
  public void close() throws Exception {
    close(Collections.newSetFromMap(new IdentityHashMap<>()), Collections.emptySet());
  }

  @SuppressWarnings({"PMD.AvoidCatchingGenericException", "PMD.SignatureDeclareThrowsException"})
  void close(Set<TelemetryBackend> closedBackends, Set<TelemetryBackend> skipBackends)
      throws Exception {
    if (skipBackends.contains(this) || !closedBackends.add(this)) {
      return;
    }

    for (Entry entry : m_entries.values()) {
      entry.close();
    }
    m_entries.clear();

    Exception exception = null;
    for (TelemetryBackend backend : m_backends) {
      try {
        TelemetryRegistry.closeBackend(backend, closedBackends, skipBackends);
      } catch (Exception e) {
        if (exception == null) {
          exception = e;
        } else {
          exception.addSuppressed(e);
        }
      }
    }
    if (exception != null) {
      throw exception;
    }
  }

  void collectOwnedBackends(Set<TelemetryBackend> ownedBackends) {
    if (!ownedBackends.add(this)) {
      return;
    }
    for (TelemetryBackend backend : m_backends) {
      TelemetryRegistry.collectOwnedBackends(backend, ownedBackends);
    }
  }

  @Override
  public TelemetryEntry getEntry(String path) {
    return m_entries.computeIfAbsent(path, k -> new Entry(m_backends, k));
  }

  @Override
  public void removeEntry(String path) {
    Entry entry = m_entries.remove(path);
    if (entry != null) {
      entry.close();
    }
    for (TelemetryBackend backend : m_backends) {
      backend.removeEntry(path);
    }
  }

  @SuppressWarnings("PMD.ForLoopCanBeForeach")
  private static final class Entry implements TelemetryEntry {
    private final List<TelemetryEntry> m_entries;
    private volatile boolean m_closed;

    Entry(List<TelemetryBackend> backends, String path) {
      List<TelemetryEntry> entries = new ArrayList<>(backends.size());
      for (TelemetryBackend backend : backends) {
        entries.add(Objects.requireNonNull(backend.getEntry(path), "backend entry"));
      }
      m_entries = List.copyOf(entries);
    }

    void close() {
      m_closed = true;
    }

    @Override
    public boolean isDiscard() {
      if (m_closed) {
        return true;
      }
      for (TelemetryEntry entry : m_entries) {
        if (!entry.isDiscard()) {
          return false;
        }
      }
      return true;
    }

    @Override
    public void keepDuplicates() {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).keepDuplicates();
      }
    }

    @Override
    public void setProperty(String key, String value) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).setProperty(key, value);
      }
    }

    @Override
    public <T> void logStruct(T value, Struct<? super T> struct, long timestamp) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).logStruct(value, struct, timestamp);
      }
    }

    @Override
    public <T> void logProtobuf(T value, Protobuf<? super T, ?> proto, long timestamp) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).logProtobuf(value, proto, timestamp);
      }
    }

    @Override
    public <T> void logStructArray(T[] value, Struct<? super T> struct, long timestamp) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).logStructArray(value, struct, timestamp);
      }
    }

    @Override
    public void logBoolean(boolean value, long timestamp) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).logBoolean(value, timestamp);
      }
    }

    @Override
    public void logByte(byte value, long timestamp) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).logByte(value, timestamp);
      }
    }

    @Override
    public void logShort(short value, long timestamp) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).logShort(value, timestamp);
      }
    }

    @Override
    public void logInt(int value, long timestamp) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).logInt(value, timestamp);
      }
    }

    @Override
    public void logLong(long value, long timestamp) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).logLong(value, timestamp);
      }
    }

    @Override
    public void logFloat(float value, long timestamp) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).logFloat(value, timestamp);
      }
    }

    @Override
    public void logDouble(double value, long timestamp) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).logDouble(value, timestamp);
      }
    }

    @Override
    public void logString(String value, String typeString, long timestamp) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).logString(value, typeString, timestamp);
      }
    }

    @Override
    public void logBooleanArray(boolean[] value, long timestamp) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).logBooleanArray(value, timestamp);
      }
    }

    @Override
    public void logShortArray(short[] value, long timestamp) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).logShortArray(value, timestamp);
      }
    }

    @Override
    public void logIntArray(int[] value, long timestamp) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).logIntArray(value, timestamp);
      }
    }

    @Override
    public void logLongArray(long[] value, long timestamp) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).logLongArray(value, timestamp);
      }
    }

    @Override
    public void logFloatArray(float[] value, long timestamp) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).logFloatArray(value, timestamp);
      }
    }

    @Override
    public void logDoubleArray(double[] value, long timestamp) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).logDoubleArray(value, timestamp);
      }
    }

    @Override
    public void logStringArray(String[] value, long timestamp) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).logStringArray(value, timestamp);
      }
    }

    @Override
    public void logRaw(byte[] value, String typeString, long timestamp) {
      if (m_closed) {
        return;
      }
      for (int i = 0; i < m_entries.size(); i++) {
        if (m_closed) {
          return;
        }
        m_entries.get(i).logRaw(value, typeString, timestamp);
      }
    }
  }
}
